test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1

```diff
@@ -0,0 +1,917 @@
+                $data.where{[system.io.path]::GetRandomFileName() -eq "Hello"}
+'@
+
+            $script2 = @'
+                $data = @(
+                    @{
+                        Node = "first"
+                        Value1 = 1
+                        Value2 = 2
+                        first = $true
+                    }
+                    @{
+                        Node = "second"
+                        Value1 = 3
+                        Value2 = 4
+                        Second = $true
+                    }
+                    @{
+                        Node = "third"
+                        Value1 = 5
+                        Value2 = 6
+                        third = $true
+                    }
+                )
+
+                $result = $data.foreach('value1')
+                Write-Output $result
+
+                # Execute method in scriptblock of foreach operator, should throw in ConstrainedLanguage mode.
+                $data.foreach{[system.io.path]::GetRandomFileName().Length}
+'@
+
+            $script3 = @'
+            # Method call should throw error.
+            (Get-Process powershell*).Foreach('GetHashCode')
+'@
+
+            $script4 = @'
+            # Where method call should throw error.
+            (get-process powershell).where{$_.GetType().FullName -match "process"}
+'@
+
+            $TestCasesForeach = @(
+                @{testName = "Verify where statement with invalid method call in constrained language is disallowed"; script = $script1 }
+                @{testName = "Verify foreach statement with invalid method call in constrained language is disallowed"; script = $script2 }
+                @{testName = "Verify foreach statement with embedded method call in constrained language is disallowed"; script = $script3 }
+                @{testName = "Verify where statement with embedded method call in constrained language is disallowed"; script = $script4 }
+            )
+        }        
+
+        It "<testName>" -TestCases $TestCasesForeach {
+
+            param (
+                [string] $script
+            )
+
+            try
+            {
+                Invoke-LanguageModeTestingSupportCmdlet -EnableConstrainedLanguageMode
+
+                $sb = [scriptblock]::Create($script)
+                & sb
+                throw "No Exception!"
+            }
+            catch
+            {
+                $exception = $_
+            }
+            finally
+            {
+                Invoke-LanguageModeTestingSupportCmdlet -EnableFullLanguageMode
+            }
+
+            $exception.FullyQualifiedErrorId | Should Match "MethodInvocationNotSupportedInConstrainedLanguage"
+        }
+    }
+
+    # End Describe blocks
+}
+finally
+{
+    if ($IsWindows)
+    {
+        Invoke-LanguageModeTestingSupportCmdlet -RevertLockdownMode -EnableFullLanguageMode
+    }
+
+    if ($defaultParamValues -ne $null)

```

Comment:
I don't think this check is needed

Reply:
I agree that it is not needed now but I prefer to keep it in case something is added to the test in the future where an error can be thrown before $defaultParamValues is assigned.  It is essential that the global variable is restored or following tests will not run correctly.